solana/solana-ibc/programs/solana-ibc/src/lib.rs

```diff
@@ -169,6 +168,67 @@ pub mod solana_ibc {
             .map_err(move |err| error!((&err)))
     }
 
+    /// Method to verify ed25519 signatures and send update client message.
+    ///
+    /// Since we cannot use ed25519 signature verification due to compute budget limit,
+    /// we have to call ed25519 program and pass the message, pubkey and signatures in
+    /// the same transaction in which the below method would be called. We then verify
+    /// the signatures comparing to the data in the previous instructions. If they
+    /// signature verification is success, we send the `UpdateClient` msg. This method
+    /// can only be called to send `UpdateClient` message. Calling otherwise would
+    /// panic.
+    pub fn send_update_client(
+        ctx: Context<SendUpdateClient>,
+        update_client_msg: ibc::MsgEnvelope,
+        signatures_data: Vec<SignatureData>,
+    ) -> Result<()> {
+        let length = signatures_data.len() as i64;
+        let mut index = 0;
+        for signature_data in signatures_data {
+            let verifier = ed25519::Verifier::new(
+                &ctx.accounts.ix_sysvar,
+                index - length,
+            )?;
+            let result = verifier.verify(
+                &signature_data.message,
+                &signature_data.pubkey.into(),
+                &signature_data.signature.into(),
+            );
+            msg!("This is the result {}", result);
+            if !result {
+                return Err(error::Error::ContextError(
+                    ibc::ClientError::HeaderVerificationFailure {
+                        reason: "Signature Verification Failed".to_string(),
+                    }
+                    .into(),
+                )
+                .into());
+            }
+            index += 1;
+        }
+        if cfg!(not(feature = "mocks")) {
+            match update_client_msg.clone() {
```

Comment:
why don't you perform this check before the signature verification?

Reply:
Because we need to check if signature verification works.